dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h

```diff
@@ -0,0 +1,234 @@
+ * as separate dimension. Instead, Width dimension is larger by `nchannels` number of times.
+ * This is because, the analysis is channel-agnostic.
+ *
+ * Function assumes, that memory layout is *HWC, and the Roi
+ * is represented as [[x_lo, y_lo], [x_hi, y_hi]].
+ * Therefore, while copying, order of values needs to be reversed.
+ */
+template <size_t ndims>
+TensorListShape<ndims> RoiToShape(const std::vector<Roi<ndims>> &rois, int nchannels) {
+  std::vector<TensorShape<ndims>> ret;
+
+  for (const auto &roi : rois) {
+    assert(all_coords(roi.hi >= roi.lo) && "Cannot create a tensor shape from an invalid Box");
+    TensorShape<ndims> ts;
+    auto e = roi.extent();
+    auto ridx = ndims;
+    ts[--ridx] = e[0] * nchannels;
+    for (size_t idx = 1; idx < ndims; idx++) {
+      ts[--ridx] = e[idx];
+    }
+    ret.emplace_back(ts);
+  }
+
+  return TensorListShape<ndims>{ret};
+}
+
+
+/**
+ * 1. If `rois` is empty, that means whole image is analysed: Roi will the the size of an image
+ * 2. If `rois` is not empty, it is assumed, that Roi is provided for every image in batch.
+ *    In this case, final Roi is an intersection of provided Roi and the image.
+ *    (This is a sanity-check for Rois, that can be larger than image)
+ */
+template <size_t spatial_dims>
+std::vector<Roi<spatial_dims>> AdjustRois(const std::vector<Roi<spatial_dims>> rois,
+                                          const TensorListShape<spatial_dims + 1> &shapes) {
+  assert(rois.empty() || rois.size() == static_cast<size_t>(shapes.num_samples()));
+  std::vector<Roi<spatial_dims>> ret(shapes.num_samples());
+
+  auto whole_image = [](auto shape) -> Roi<spatial_dims> {
+      ivec<spatial_dims> size;
+      for (size_t i = 0; i < spatial_dims; i++)
+        size[i] = shape[spatial_dims - 1 - i];
+      return {0, size};
+  };
+
+  if (rois.empty()) {
+    for (int i = 0; i < shapes.num_samples(); i++) {
+      ret[i] = whole_image(shapes[i]);
+    }
+  } else {
+    for (size_t i = 0; i < rois.size(); i++) {
+      ret[i] = intersection(rois[i], whole_image(shapes[i]));
+    }
+  }
+
+  return ret;
+}
+
+
+/**
+ * Note: Since the brightness-contrast calculation is channel-agnostic (it is performed in the same
+ * way for every channel), SampleDescriptor assumes, that sample is channel-agnostic. Therefore it
+ * needs to be flattened
+ * @see RoiToShape
+ */
+template <class OutputType, class InputType, int ndims>
+std::vector<SampleDescriptor<OutputType, InputType, ndims - 1>>
+CreateSampleDescriptors(const OutListGPU<OutputType, ndims> &out,
+                        const InListGPU<InputType, ndims> &in,
+
+                        const std::vector<float> &brightness, const std::vector<float> &contrast) {
+  std::vector<SampleDescriptor<OutputType, InputType, ndims - 1>> ret(in.num_samples());
+
+  for (int i = 0; i < in.num_samples(); i++) {
+    auto &sample = ret[i];
+    sample.in = in[i].data;
+    sample.out = out[i].data;
+    sample.in_pitch = {};
```

Comment:
Aren't those already initialized by the vector constructor? Also you rewrite them just a few lines below, maybe it's ok to get rid of those two lines. Or, as this is a POD, maybe 

```suggestion
    sample = {in[i].data, out[i].data, flattened_pitch(in[i]), flattened_pitch(out[i]), brightness[i], contrast[i]};
```

Reply:
I prefer my way, cause yours will be messed up, if somebody changes the order of fields in struct...

Reply:
Sure, no strong opinion about it.